Add proposal for partial type inference (#7582)
dotnet/csharplang: proposals/partial-type-inference.md

# Partial type inference

> Note: This proposal was created because of championed [Partial type inference](https://github.com/dotnet/csharplang/issues/1349). It is a continuation of the proposed second version published in [csharplang/discussions/7467](https://github.com/dotnet/csharplang/discussions/7467) and the first version published in [csharplang/discussions/7286](https://github.com/dotnet/csharplang/discussions/7286)

> ## Summary

Reviewer: Chatted with @jaredpar about the whole proposal. There's three parts that I can see:
1. allowing free/floating type arguments in method invocation (`M<int, _>(...)`)
2. allowing free/floating type arguments in object creation **with floating arity** (`new C<>(...)`)
3. allowing object initializers contribute to the inference in object creation (`new C<>(...) { ...initializer... }`)

The second and third part seem much more complicated as they introduce bidirectional type inference.

What we would propose instead as core design proposal is:
1. allowing free/floating type arguments in method invocation [KEEP]
2. allowing free/floating type arguments in object creation **with fixed arity** using the same syntax: `new C<_, _>()`
The benefit of restricting object creation to fixed/known arity is that the flow for object creation remains more like today: pick a single type and apply overload resolution to its constructors (and overload resolution can now deal with fixed vs. floating type arguments).

Those two core parts would be very useful, straightforward and not very controversial with LDM.

We would recommend either removing or clearly separating out (potentially in a separate doc) the two parts that are controversial and complex:
1. allowing floating arity in object creation
2. allowing initializers to contribute to type inference in object creation

Author: Should by fixed by commit [ea87f2f](https://github.com/dotnet/csharplang/pull/7582/commits/ea87f2f973aff156f9467c56d4e1861e3c7306bb). I moved the mentioned parts (initializers and floating arity) to alternatives.

  Concretely, we can see the weakness in cases where type arguments only depend on target type or type parameter restrictions.

  > Example
  >
  > ```csharp
  > object data = database.fetch();
  > int count = data.Field("count"); // Error, method type inference can't infer the return type. We have to specify the type argument.
  > 
  > public static class Extensions
  > {
  >     public static TReturn Field<TReturn>(this object inst, string fieldName) { ... }
  > }
  > ```
  >
  > ```csharp
  > test(new MyData()); // Error, method type inference can't infer T. We have to specify all type arguments.
  >
  > public void test<T, U>(U data) where T : TestCaseDefault<U> { ... }
  > ```

  The second improvement regards the "all or nothing" principle, where the method type inference infers either all of the type arguments or nothing.

  > Example
  > ```csharp
  > log<, object>(new Message( ... ), null); // Error, impossible to specify just one type argument. We have to specify all of them. 
  >
  > public void log<T, U>(T message, U appendix) { ... }
  > ```

>   The first improvement, which would improve the method type inference algorithm, has a significant disadvantage of the breaking change.

Reviewer 2: ```suggestion
  The first improvement, which would improve the method type inference algorithm, has a significant disadvantage of introducing a breaking change.
```

>   On the other hand, the second improvement, which would enable specifying some of the method's type arguments, does not influence old code, solves problems regarding the "all or nothing" principle, and reduces the first weakness.

Reviewer 2: I'm unclear what the "all or nothing" principle means here.

Author: Line [58](https://github.com/dotnet/csharplang/pull/7582/files#diff-98ed61744ec680992e9779b8a70429db7e853f6942a0cb4e94b43b8e7ae96b91R58) explains "all or nothing" principle. The method type inference infers either all of the type arguments or nothing (e.g. we can't hint the compiler some of the type arguments).

Reviewer 3: In line [68](https://github.com/dotnet/csharplang/pull/7582/files/7aa277e63f69c12f3bd27cae6c98a7a965fa6958#diff-98ed61744ec680992e9779b8a70429db7e853f6942a0cb4e94b43b8e7ae96b91R68), wouldn't it even be possible to infer twice: `test<TestCaseDefault<_>, _>(new MyData());` ?

- **IDE** - Improvement of the type inference can complicate IDE hints during coding. 
We should give the user clear and not overwhelming errors when there will be an error and try to provide info that helps him to fix it.
- **Extensions** - We don't want to make this change blocker for another potential feature in the future. 
So will want to look ahead to other potential directions, which can be done after this feature.

## Detailed design
[design]: #detailed-design

### Grammar

> Specification: Original section changed in the following way

The following changes are made in [tokens](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#64-tokens) located in the [grammar](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#62-grammars) section.

> [Identifiers](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#643-identifiers)

- The semantics of an identifier named `_` depends on the context in which it appears:
  - It can denote a named program element, such as a variable, class, or method, or
  - It can denote a discard (§9.2.9.1).

>   - \***It can denote an inferred type argument avoiding specifying type arguments which can be inferred by the compiler.**

Reviewer 2: Rather than _it can_ I'd prefer the language _it will_. Essentially when flipping to `/langversion:X` where `X` is the release that introduces this feature, I'd prefer the meaning to be unambiguous. That is for the case `M<_, object>(42, null)`, the `_` unambiguously means an inferred type parameter. 

Yes this would be breaking change to developers who defined `class _ { }` but that seems like a minority case. It in the same ballpark as `class record` which we broke when introducing records.

Author: Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)

  > ```   

* A method group and type are said to be *partial_inferred* if it contains at least one *inferred_type_argument*. 

* A type is said to be *generic_inferred* when all the following hold:
  * It has an empty *type_argument_list*.
  * It occurs as a *type* of *object_creation_expression*.
  > Example
  >
  > ```csharp
  > new C<>(...) // Valid code, C is generic_inferred.
  > new C<G<>>(...) // Invalid code, C nor G are generic_inferred.
  > F<>(...) // Invalid code, F isn't generic_inferred.
  > ```

### Namespace and type names

> Specification: Original section changed in the following way

Determining the [meaning](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/basic-concepts.md#781-general) of a *namespace_or_type_name* is changed as follow.

* If a type is a *generic_inferred*, then we resolve the identifier in the same manner except ignoring the arity of type parameters (Types of arity 0 is ignored). 
If there is an ambiguity in the current scope, a compilation-time error occurs.
  > Example
  >
  > ```csharp
  > class P1
  > {
  >     void M() 
  >     {

>   >         new C1<>( ... ); // Refers generic_inferred type C1<T>

Reviewer: `T`, `T1` and `T2` don't exist here.
Reading ahead to the "Object creation expressions" section, I believe the comments in this snippet are not what you intended and some actual types should be inferred from inspecting candidate constructors.

Author: Should be fixed by commit [b6081e3](https://github.com/dotnet/csharplang/pull/7582/commits/b6081e37e9a7e3b6892d3b0164e5d91b3a61cfc3)

  - **Type inference ([§12.6.3](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1263-type-inference)) succeeds, inferring a list of *inferred_type_arguments* for the call, and**
  - **Once the *inferred_type_arguments* are inferred and together with remaining type arguments are substituted for the corresponding method type parameters, all constructed types in the parameter list of `F` satisfy their constraints ([§8.4.5](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/types.md#845-satisfying-constraints)), and the parameter list of `F` is applicable with respect to `A` ([§12.6.4.2](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#12642-applicable-function-member))**
- If `F` is generic and `M` includes a type argument list, `F` is a candidate when:
  - `F` has the same number of method type parameters as were supplied in the type argument list, and
  - Once the type arguments are substituted for the corresponding method type parameters, all constructed types in the parameter list of `F` satisfy their constraints (§8.4.5), and the parameter list of `F` is applicable with respect to `A` (§12.6.4.2).

> ### Object creation expressions

Reviewer: I stopped here for today. Will resume tomorrow

# Partial type inference

> * [x] Proposed

Reviewer: Those bullets are best kept in the championed proposal ([https://github.com/dotnet/csharplang/issues/1349](https://github.com/dotnet/csharplang/issues/1349)) so they can be updated fluidly

Reviewer: I made some edits directly to save time. I hope that's okay with you.
I removed these bullets and simplified the links in the copied sections of the spec (makes it easier to review the diff).

* [ ] Prototype
* [ ] Implementation
* [ ] Specification

> Note: This proposal was created because of championed [Partial type inference](https://github.com/dotnet/csharplang/issues/1349). It is a continuation of the proposed second version published in [csharplang/discussions/7467](https://github.com/dotnet/csharplang/discussions/7467) and the first version published in [csharplang/discussions/7286](https://github.com/dotnet/csharplang/discussions/7286)

## Summary
[summary]: #summary

> Partial type inference introduces a syntax skipping obvious type arguments in the argument list of

Reviewer: >obvious [](http://example.com/codeflow?start=52&length=7)

nit: instead of "obvious" consider "inferrable" (also applies below)

Author: Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)

  >
  > public void log<T, U>(T message, U appendix) { ... }
  > ```

  The first improvement, which would improve the method type inference algorithm, has a significant disadvantage of the breaking change.
  On the other hand, the second improvement, which would enable specifying some of the method's type arguments, does not influence old code, solves problems regarding the "all or nothing" principle, and reduces the first weakness.
  
  > Example
  > 
  > ```csharp
  > test<TestCaseDefault<MyData>, _>(new MyData()); // We can use _ to mark type arguments which should be inferred by the compiler.
  >
  > public void test<T, U>(U data) where T : TestCaseDefault<U> { ... }
  > ```
  >
  > ```csharp
  > log<_, object>(new Message( ... ), null); // We can use _ to mark type arguments which should be inferred by the compiler.
  >
  > public void log<T, U>(T message, U appendix) { ... }
  > ```

* The next motivation is constructor type inference. 
  Method type inference is not defined on *object_creation_expression*, prohibiting taking advantage of type inference.
  We divide use cases into the following categories, where type inference would help the programmer. 

>   1. Cases where the method type inference would success.

Reviewer: ```suggestion
  1. Cases where the method type inference would succeed.
```

   
  > Example
  > 
  > ```csharp

>   > var wrappedData = Create(new MyData());

Reviewer: nit: Do we need this line to illustrate the point? 
I think the line below (`return new Wrapper<T>(item);`) is central and sufficient

Author: Should be fixed by commit [5e59e96](https://github.com/dotnet/csharplang/pull/7582/commits/5e59e96d98c7c7c12678c293b04d5e20d8b8f037)

  > 
  > public static Wrapper<T> Create<T>(T item) { return new Wrapper<T>(item); }
  > 
  > class Wrapper<T> { public Wrapper(T item) { ... } }
  > ```
  

>   2. Cases where the method type inference would be weak. (Using type info from target type, or type arguments' constrains)

Reviewer: ```suggestion
  2. Cases where the method type inference would be weak. (Using type info from target type, or type constraints)
```

typo: constrains -> constraints, also applies in other locations below

Author: Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)

  > ```
  > 
  > ```csharp
  > new Dictionary<_, string>() {[""] = null};
  > ```

No matter how the partial type inference would work, we should be careful about the following things.

- **Convenience** - We want an easy and intuitive syntax that we can skip the obvious type arguments.
- **Performance** - Type inference is a complicated problem when we introduce subtyping and overloading in a type system.
Although it can be done, the computation can take exponential time which we don't want.
So it has to be restricted to cases, where the problem can be solved effectively but it still has practical usage.
- **IDE** - Improvement of the type inference can complicate IDE hints during coding. 
We should give the user clear and not overwhelming errors when there will be an error and try to provide info that helps him to fix it.
- **Extensions** - We don't want to make this change blocker for another potential feature in the future. 
So will want to look ahead to other potential directions, which can be done after this feature.

## Detailed design
[design]: #detailed-design

### Grammar

> Specification: Original section changed in the following way

> The following changes are made in [tokens](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#64-tokens) located in the [grammar](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#62-grammars) section.

Reviewer: nit: Can be simplified
```suggestion
We modify [Identifiers](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#643-identifiers) as follows:
```
(then the "Identifiers" link below becomes redundant)

> [Identifiers](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#643-identifiers)

- The semantics of an identifier named `_` depends on the context in which it appears:
  - It can denote a named program element, such as a variable, class, or method, or
  - It can denote a discard ([§9.2.9.1](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/variables.md#9291-discards)).

>   - **It can denote an inferred type argument avoiding specifying type arguments which can be inferred by the compiler.**

Reviewer: Can be simplified
```suggestion
  - **It can denote a type argument to be inferred.**
```

Author: Should be fixed by commit [33b71f0](https://github.com/dotnet/csharplang/pull/7582/commits/33b71f0c489eaa03018d18a9cabf8b2cdf74a43a)

> [Keywords](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/lexical-structure.md#644-keywords)

* A ***contextual keyword*** is an identifier-like sequence of characters that has special meaning in certain contexts, but is not reserved, and can be used as an identifier outside of those contexts as well as when prefaced by the `@` character.

  ```diff
  contextual_keyword
      : 'add'    | 'alias'      | 'ascending' | 'async'     | 'await'
      | 'by'     | 'descending' | 'dynamic'   | 'equals'    | 'from'
      | 'get'    | 'global'     | 'group'     | 'into'      | 'join'
      | 'let'    | 'nameof'     | 'on'        | 'orderby'   | 'partial'
      | 'remove' | 'select'     | 'set'       | 'unmanaged' | 'value'
  +   | 'var'    | 'when'       | 'where'     | 'yield'     | '_'
  -   | 'var'    | 'when'       | 'where'     | 'yield'
    ;
  ```

### Type arguments

> Specification: Original section changed in the following way

* We change the meaning of the content of *type_argument_list* in two contexts.

>   * [Constructed types](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/types.md#84-constructed-types) occuring in [*object_creation_expression*](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#128162-object-creation-expressions)

Reviewer: As a general comment, I appreciate the references back to the standard/spec. The format you adopted in other sections (copy a paragraph from original spec and add some highlighted edits) is really nice. But when I look at "Type arguments" section, it's not clear which section of the spec is being edited. 

Maybe you mean to edit section [8.4.2](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#842-type-arguments) ("type arguments")? That seems to be the right place to introduce ***inferred_type_argument*** into the ***type_argument_list*** grammar.
(If so, then the references to "constructed types" (8.4) and  "object_creation_expression" could be removed)

Below is a possible grammar change to introduce `inferred_type_argument`:
  ```antlr
type_argument_list
    : '<' type_arguments '>'
    ;
    
type_arguments // modified
    : type_argument (',' type_arguments)
    | inferred_type_argument (',' type_arguments)
    ;   

type_argument
    : type
    ;

inferred_type_argument // new
    : '_'
    ;
  ```
Then we can say that ***inferred_type_argument*** is only allowed in the contexts that you listed.

We still need to resolve the issue of back compat:
```
public class _ { }
public class C<T> 
{
    public void M(C<_> x) { } // allowed today
}
```
We'll need to decide whether the presence of an existing type named "_"  in scope affects the interpretation of "_" in type argument lists, or whether "_" always means "inferred type argument" (ie. we take a compat break, maybe warn on types named "_").

Author: Should be fixed by [6039b9f](https://github.com/dotnet/csharplang/pull/7582/commits/6039b9fcc71f26a254f65ce4261200d04ba90e17).

1. *invocation_expression*
2. *object_creation_expresssion*

and allowing to specify just ambiguous ones.

> Example of skipping inferrable type arguments
>
> ```csharp
> M<_, object>(42, null); 
> void M<T1, T2>(T1 t1, T2 t2) { ... }
> ```

It also improves the type inference in the case of *object_creation_expression* by leveraging type bounds obtained from the target and *type_parameter_constraints_clauses*. 

> Example of inferring type arguments of type from the initializer list
> 
> ```csharp
> using System.Collections.Generic;
> IList<int> temp = new List<_>();
> ```

Besides the changes described above, the proposal mentions further interactions and possibilities to extend the partial type inference.

## Motivation
[motivation]: #motivation

* The current method type inference has two possible areas for improvement.
  The first improvement regards the strength of the method type inference, which uses only arguments' types to deduce the method's type arguments.
  Concretely, we can see the weakness in cases where type arguments only depend on target type or type parameter restrictions.

  > Example
  >
  > ```csharp
  > object data = database.fetch();

>   > int count = data.Field("count"); // Error, method type inference can't infer the return type. We have to specify the type argument.

Author: Should be fixed by commit [e973d80](https://github.com/dotnet/csharplang/pull/7582/commits/e973d80d053e19997a8dde270eb8579e8d167a82)

      : 'add'    | 'alias'      | 'ascending' | 'async'     | 'await'
      | 'by'     | 'descending' | 'dynamic'   | 'equals'    | 'from'
      | 'get'    | 'global'     | 'group'     | 'into'      | 'join'
      | 'let'    | 'nameof'     | 'on'        | 'orderby'   | 'partial'
      | 'remove' | 'select'     | 'set'       | 'unmanaged' | 'value'
  +   | 'var'    | 'when'       | 'where'     | 'yield'     | '_'
  -   | 'var'    | 'when'       | 'where'     | 'yield'
    ;
  ```

### Type arguments

> Specification: Original section changed in the following way

> We change [type arguments](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/types.md#842-type-arguments) section as follows:

* ***inferred_type_argument*** represents an unknown type, which will be resolved during type inference. 

* Each argument in a type argument list is ~~simply a type~~ **a type or an inferred type**.

```diff
type_argument_list
  : '<' type_arguments '>'
  ;
  
type_arguments
-  : type_argument (',' type_argument)*
+  : type_argument (',' type_arguments)
+  | inferred_type_argument (',' type_arguments)
  ;   

type_argument
  : type
  ;

+inferred_type_argument
+  : '_'
+  ;
```

> * When a type with `_` identifier is presented in the scope where **inferred_type_argument** is used, a warning should appear since the **inferred_type_argument** hides the type's name causing a breaking change.

Author: Should be fixed by commit [4485ab9](https://github.com/dotnet/csharplang/pull/7582/commits/4485ab9ea0bdb6a4c0b04c3575728caf54e1a19d)

  - **A binding-time error occurs when:**
    - **The set of candidate instance constructors is empty, or if a single best instance constructor cannot be identified, and there is no chance to know the target type now or later.**
  - **If the set of candidate instance constructors is still empty, or if a single best instance constructor cannot be identified, repeat the binding of the *object_creation_expression* to the time, when target type will be known and add it to inputs of type inference.**
  - The result of the *object_creation_expression* is a value of type `T`, namely the value produced by invoking the instance constructor determined in the two steps above.
  - Otherwise, the *object_creation_expression* is invalid, and a binding-time error occurs.

### Type inference

> Specification: Original section changed in the following way

> We change the [type inference](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1263-type-inference) as follows.

Author: Should by fixed by commit [9dfa24f](https://github.com/dotnet/csharplang/pull/7582/commits/9dfa24f073118a81555a11db2edda8de4cc6846c)

  * If the inferred type contains a nonempty *type_argument_list*.
    * We replace each `_` identifier with a new type variable `X`.
    * We perform *shape inference* from each type argument to the corresponding type parameter.
  * If the target type should be used based on the expression binding, perform *upper-bound inference* from it to the type containing the constructor
  * If the expression contains *where* clauses defining type constraints of type parameters of the type containing constructor, for each constraint not representing *constructor* constraint, *reference type constraint*, *value type constraint* and *unmanaged type constraint* perform *lower-bound inference* from the constraint to the corresponding type parameter.

* Arguments binding
  * It can happen that an argument of an expression will be *object_creation_expression*, which needs a target type to be successful binded. 
  * In these situations, we behave like the type of the argument is unknown and bind it when we will know the target type.
  * We treat it in the same manner as an unconverted *new()* operator.

#### Type inference algorithm change

> > Specification: Original section changed in the following way.

Author: Should be fixed by commit [7aa277e](https://github.com/dotnet/csharplang/pull/7582/commits/7aa277e63f69c12f3bd27cae6c98a7a965fa6958)

### Compile-time checking of dynamic member invocation

We change the [compile-time checking](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/expressions.md#1265-compile-time-checking-of-dynamic-member-invocation) in order to be useful during partial type inferece.

- First, if `F` is a generic method and type arguments were provided, then those ***, that aren't *inferred_type_argument*** are substituted for the type parameters in the parameter list. However, if type arguments were not provided, no such substitution happens.
- Then, any parameter whose type is open (i.e., contains a type parameter; see §8.4.3) is elided, along with its corresponding parameter(s).
- **If the method group contains inferred type arguments, a warning should appear since partial type inference is not supported by runtime.** 

We add the following
- If an object creation expression is inferred and the argument list contains a dynamic value, an error should appear since constructor type inference is not supported by runtime.  

### Nullability

We add the following section about nullability. 

> * We can use an examination mark `?` to say that the inferred type argument should be a nullable type (e.g. `F<_?>(...)`).

Reviewer 4: ```suggestion
* We can use a question mark `?` to say that the inferred type argument should be a nullable type (e.g. `F<_?>(...)`).
```
Also, if this is intended to update a particular section of the language standard, let's link to that section here.

Author: Solved by commit [98d835e](https://github.com/dotnet/csharplang/pull/7582/commits/98d835e3dcbd67efbb7ae1c87cf54ac19578d68f)
